src/repo.rs

```diff
@@ -108,6 +110,92 @@ impl RepoHandle {
             .expect("Failed to send repo event.");
         handle
     }
+
+    pub async fn connect_remote<RecvErr, SendErr, S>(
+        &self,
+        repo_id: RepoId,
+        stream: S,
+    ) -> Result<(), error::RunConnection<SendErr, RecvErr>>
+    where
+        RecvErr: std::fmt::Debug,
+        SendErr: std::error::Error,
+        S: Sink<NetworkMessage, Error = SendErr> + Stream<Item = Result<NetworkMessage, RecvErr>>,
+    {
+        let (mut sink, stream) = stream.split();
+        let outbox = self
+            .outboxes
+            .lock()
+            .entry(repo_id)
+            .or_insert_with(|| Arc::new(Mutex::new(Default::default())))
+            .clone();
+        let inbox = self
+            .inboxes
+            .lock()
+            .entry(repo_id)
+            .or_insert_with(|| Arc::new(Mutex::new(Default::default())))
+            .clone();
+        let sending = OutgoingStream(outbox)
+            .map(|m| Ok::<_, SendErr>(m))
+            .forward(&mut sink)
+            .fuse();
+
+        let recv_msg = self.repo_sender.clone();
+        let recving = stream
+            .map_err(error::RunConnection::<SendErr, RecvErr>::Recv)
+            .try_for_each({
+                move |m| {
+                    inbox.lock().messages.push_back(m);
+                    let _ = recv_msg.try_send(RepoEvent::IncomingSyncMessage);
```

Comment:
Seems like a good idea, for now I am just trying to get it to work.

Reply:
Also, in some way this is an argument for changing the connect stream concept, because spanreed cannot know when the caller needs to slow down. In the above there is a bounded channel, which I guess could be user-configured, but at the end of the day it  seems to me that it is rather the caller--the networking app--that would know how and when to apply backpressure. 

The advantage of the code in `main` is that the repo polls the sink and streams for readiness, and it's entirely up to the provider of these to deal with internal buffering and backpressure(the repo remains a "dumb" poller). 

In this setup, while accepting a stream/sink in `connect_repo`, the repo also provides it's own stream and sink to connect those to the backend, which suddenly makes the repo responsible for buffering with the in- and outbox, and the polling is done by the outside runtime.

Reply:
That's a fair argument I think. My main thinking with the connect stream stuff was that we should provide an API for the user to just pass a stream to spanreed, because that matches more closely how the async ecosystem works. If simple backpressure can be achieved more readily by polling the stream but we can achieve that whilst still offering the stream-per-connection API then maybe that's the route we should go down.

Reply:
An alternative to manually polling the streams might be to use an async channel implementation such as [thingbuf](https://docs.rs/thingbuf/0.1.4/thingbuf/mpsc/index.html) instead of the crossbeam channel. That way in `try_send` we can use the async send operation to wait for spanreed to have pulled a message out of the channel. I think this naturally pushes the backpressure out to the application without needing to manually poll the streams?